schema/elements.go

```diff
@@ -250,14 +257,47 @@ func (s *Schema) FindNamedType(name string) (TypeDef, bool) {
 // This allows callers to not care about the difference between a (possibly
 // inlined) reference and a definition.
 func (s *Schema) Resolve(tr TypeRef) (Atom, bool) {
+	result := Atom{}
+
 	if tr.NamedType != nil {
 		t, ok := s.FindNamedType(*tr.NamedType)
 		if !ok {
 			return Atom{}, false
 		}
-		return t.Atom, true
+
+		result = t.Atom
+	} else {
+		result = tr.Inlined
 	}
-	return tr.Inlined, true
+
+	// Allow field-level electives to override the referred type's modifiers
+	switch {
+	case result.Map != nil:
+		mapCopy := *result.Map
```

Comment:
Won't this make a lot of garbage? Can we at least do the if statement first and only do that if needed?

Reply:
yes I was conflicted about this vs caching the resolved responses in a map as a field of `Schema` vs something else.

I have changed the code to have the allocations be conditional

Reply:
Looks good now, can you run the benchmarks to see if this is too bad? (might have to edit something to make sure it's exercising this?)

Reply:
I wrote up a benchmark for this implementation, which showed to me a 100% cost in time and allocations when the `elementRelationship` override was used for ALL fields. This to me was too large a cost, so I refactored this PR to cache the result when `TypeRef` included overrides.

I thought it would be a reasonable conservative benchmark to randomly instrument 10% of ALL usages of `namedType` within fields with `elementRelationship: atomic`, and also show with 100% of ALL usages. Below is the results with the current state of the PR:

<details>
<summary>10% Fields Exercising Feature</summary>

```
name                                                  time/op
Operations/Pod/Create-20                              50.2µs ± 0%
Operations/Pod/Apply-20                                123µs ± 1%
Operations/Pod/ApplyTwice-20                           281µs ± 1%
Operations/Pod/Update-20                               118µs ± 0%
Operations/Pod/UpdateVersion-20                        166µs ± 1%
Operations/Node/Create-20                             71.7µs ± 1%
Operations/Node/Apply-20                               183µs ± 1%
Operations/Node/ApplyTwice-20                          434µs ± 1%
Operations/Node/Update-20                              185µs ± 1%
Operations/Node/UpdateVersion-20                       268µs ± 1%
Operations/Endpoints/Create-20                        5.90µs ± 1%
Operations/Endpoints/Apply-20                         12.5µs ± 0%
Operations/Endpoints/ApplyTwice-20                     649µs ± 0%
Operations/Endpoints/Update-20                         630µs ± 1%
Operations/Endpoints/UpdateVersion-20                 1.24ms ± 1%
Operations/CustomResourceDefinition/Create-20          598µs ± 0%
Operations/CustomResourceDefinition/Apply-20          1.41ms ± 0%
Operations/CustomResourceDefinition/ApplyTwice-20     3.00ms ± 0%
Operations/CustomResourceDefinition/Update-20         1.30ms ± 0%
Operations/CustomResourceDefinition/UpdateVersion-20  1.94ms ± 0%

name                                                  alloc/op
Operations/Pod/Create-20                              21.0kB ± 0%
Operations/Pod/Apply-20                               52.5kB ± 0%
Operations/Pod/ApplyTwice-20                           106kB ± 0%
Operations/Pod/Update-20                              39.4kB ± 0%
Operations/Pod/UpdateVersion-20                       51.7kB ± 0%
Operations/Node/Create-20                             30.0kB ± 0%
Operations/Node/Apply-20                              73.6kB ± 0%
Operations/Node/ApplyTwice-20                          152kB ± 0%
Operations/Node/Update-20                             60.3kB ± 0%
Operations/Node/UpdateVersion-20                      82.0kB ± 0%
Operations/Endpoints/Create-20                        3.55kB ± 0%
Operations/Endpoints/Apply-20                         6.41kB ± 0%
Operations/Endpoints/ApplyTwice-20                     126kB ± 0%
Operations/Endpoints/Update-20                         119kB ± 0%
Operations/Endpoints/UpdateVersion-20                  233kB ± 0%
Operations/CustomResourceDefinition/Create-20          170kB ± 0%
Operations/CustomResourceDefinition/Apply-20           483kB ± 0%
Operations/CustomResourceDefinition/ApplyTwice-20      962kB ± 0%
Operations/CustomResourceDefinition/Update-20          335kB ± 0%
Operations/CustomResourceDefinition/UpdateVersion-20   455kB ± 0%

name                                                  allocs/op
Operations/Pod/Create-20                                 473 ± 0%
Operations/Pod/Apply-20                                1.25k ± 0%
Operations/Pod/ApplyTwice-20                           2.61k ± 0%
Operations/Pod/Update-20                                 970 ± 0%
Operations/Pod/UpdateVersion-20                        1.40k ± 0%
Operations/Node/Create-20                                557 ± 0%
Operations/Node/Apply-20                               1.56k ± 0%
Operations/Node/ApplyTwice-20                          3.51k ± 0%
Operations/Node/Update-20                              1.41k ± 0%
Operations/Node/UpdateVersion-20                       2.16k ± 0%
Operations/Endpoints/Create-20                          76.0 ± 0%
Operations/Endpoints/Apply-20                            148 ± 0%
Operations/Endpoints/ApplyTwice-20                     6.33k ± 0%
Operations/Endpoints/Update-20                         6.17k ± 0%
Operations/Endpoints/UpdateVersion-20                  12.2k ± 0%
Operations/CustomResourceDefinition/Create-20          3.71k ± 0%
Operations/CustomResourceDefinition/Apply-20           10.3k ± 0%
Operations/CustomResourceDefinition/ApplyTwice-20      21.4k ± 0%
Operations/CustomResourceDefinition/Update-20          8.26k ± 0%
Operations/CustomResourceDefinition/UpdateVersion-20   12.7k ± 0%
```
</details>

<details>
<summary>100% of fields exercising feature</summary>

```
name                                                  time/op
Operations/Pod/Create-20                              50.3µs ± 0%
Operations/Pod/Apply-20                                123µs ± 1%
Operations/Pod/ApplyTwice-20                           281µs ± 1%
Operations/Pod/Update-20                               118µs ± 1%
Operations/Pod/UpdateVersion-20                        166µs ± 1%
Operations/Node/Create-20                             72.0µs ± 1%
Operations/Node/Apply-20                               184µs ± 1%
Operations/Node/ApplyTwice-20                          437µs ± 1%
Operations/Node/Update-20                              186µs ± 0%
Operations/Node/UpdateVersion-20                       268µs ± 0%
Operations/Endpoints/Create-20                        5.88µs ± 2%
Operations/Endpoints/Apply-20                         12.5µs ± 0%
Operations/Endpoints/ApplyTwice-20                     648µs ± 0%
Operations/Endpoints/Update-20                         628µs ± 1%
Operations/Endpoints/UpdateVersion-20                 1.24ms ± 1%
Operations/CustomResourceDefinition/Create-20          698µs ±14%
Operations/CustomResourceDefinition/Apply-20          1.62ms ±13%
Operations/CustomResourceDefinition/ApplyTwice-20     3.30ms ± 9%
Operations/CustomResourceDefinition/Update-20         1.43ms ± 9%
Operations/CustomResourceDefinition/UpdateVersion-20  2.05ms ± 5%

name                                                  alloc/op
Operations/Pod/Create-20                              21.0kB ± 0%
Operations/Pod/Apply-20                               52.5kB ± 0%
Operations/Pod/ApplyTwice-20                           106kB ± 0%
Operations/Pod/Update-20                              39.4kB ± 0%
Operations/Pod/UpdateVersion-20                       51.7kB ± 0%
Operations/Node/Create-20                             30.0kB ± 0%
Operations/Node/Apply-20                              73.6kB ± 0%
Operations/Node/ApplyTwice-20                          152kB ± 0%
Operations/Node/Update-20                             60.3kB ± 0%
Operations/Node/UpdateVersion-20                      82.0kB ± 0%
Operations/Endpoints/Create-20                        3.55kB ± 0%
Operations/Endpoints/Apply-20                         6.41kB ± 0%
Operations/Endpoints/ApplyTwice-20                     126kB ± 0%
Operations/Endpoints/Update-20                         119kB ± 0%
Operations/Endpoints/UpdateVersion-20                  233kB ± 0%
Operations/CustomResourceDefinition/Create-20          210kB ±19%
Operations/CustomResourceDefinition/Apply-20           574kB ±16%
Operations/CustomResourceDefinition/ApplyTwice-20     1.08MB ±11%
Operations/CustomResourceDefinition/Update-20          386kB ±13%
Operations/CustomResourceDefinition/UpdateVersion-20   509kB ±11%

name                                                  allocs/op
Operations/Pod/Create-20                                 473 ± 0%
Operations/Pod/Apply-20                                1.25k ± 0%
Operations/Pod/ApplyTwice-20                           2.61k ± 0%
Operations/Pod/Update-20                                 970 ± 0%
Operations/Pod/UpdateVersion-20                        1.40k ± 0%
Operations/Node/Create-20                                557 ± 0%
Operations/Node/Apply-20                               1.56k ± 0%
Operations/Node/ApplyTwice-20                          3.51k ± 0%
Operations/Node/Update-20                              1.41k ± 0%
Operations/Node/UpdateVersion-20                       2.16k ± 0%
Operations/Endpoints/Create-20                          76.0 ± 0%
Operations/Endpoints/Apply-20                            148 ± 0%
Operations/Endpoints/ApplyTwice-20                     6.33k ± 0%
Operations/Endpoints/Update-20                         6.17k ± 0%
Operations/Endpoints/UpdateVersion-20                  12.2k ± 0%
Operations/CustomResourceDefinition/Create-20          4.43k ±16%
Operations/CustomResourceDefinition/Apply-20           12.0k ±14%
Operations/CustomResourceDefinition/ApplyTwice-20      23.5k ± 9%
Operations/CustomResourceDefinition/Update-20          9.09k ± 9%
Operations/CustomResourceDefinition/UpdateVersion-20   13.6k ± 7%
```

The CRDs example seems especially impacted with 100% fields annotated because the `CustomResourceDefinition` type and its nested subtypes like (JSONSchemaProps) have a large amount of `namedType` usages which would recursively create extra instances in this example.
</details>

Reply:
You've instrumented the code in two different ways to run these two separate benchmark, can you add a separate test instead so that we can commit both versions?

Reply:
latest commits add these benchmarks to `real_test.go`